public/app/plugins/panel/datagrid/state.ts

```diff
@@ -0,0 +1,299 @@
+
+import {
+  DatagridContextMenuData,
+  DEFAULT_CONTEXT_MENU,
+  DEFAULT_RENAME_INPUT_DATA,
+  EMPTY_GRID_SELECTION,
+  getCellWidth,
+  isDatagridEditEnabled,
+  RenameColumnInputData,
+} from './utils';
+
+interface DatagridState {
+  columns: SizedGridColumn[];
+  contextMenuData: DatagridContextMenuData;
+  renameColumnInputData: RenameColumnInputData;
+  gridSelection: GridSelection;
+  columnFreezeIndex: number;
+  toggleSearch: boolean;
+  isResizeInProgress: boolean;
+}
+
+interface UpdateColumnsPayload {
+  frame: DataFrame;
+}
+
+interface ColumnResizeStartPayload {
+  columnIndex: number;
+  width: number;
+}
+
+interface OpenCellContextMenuPayload {
+  event: CellClickedEventArgs;
+  cell: Item;
+}
+
+interface OpenHeaderContextMenuPayload {
+  event: HeaderClickedEventArgs;
+  columnIndex: number;
+}
+
+interface OpenHeaderDropdownMenuPayload {
+  screenPosition: Rectangle;
+  columnIndex: number;
+  value: string;
+}
+
+interface ColumnMovePayload {
+  from: number;
+  to: number;
+}
+
+interface MultipleCellsSelectedPayload {
+  selection: GridSelection;
+}
+
+interface ColumnFreezeChangedPayload {
+  columnIndex: number;
+}
+
+export interface DatagridAction {
+  type: DatagridActionType;
+  payload?:
+    | UpdateColumnsPayload
+    | ColumnResizeStartPayload
+    | OpenCellContextMenuPayload
+    | OpenHeaderContextMenuPayload
+    | OpenHeaderDropdownMenuPayload
+    | ColumnMovePayload
+    | MultipleCellsSelectedPayload
+    | ColumnFreezeChangedPayload;
+}
+
+export enum DatagridActionType {
+  columnResizeStart = 'columnResizeStart',
+  columnResizeEnd = 'columnResizeEnd',
+  columnMove = 'columnMove',
+  updateColumns = 'updateColumns',
+  showColumnRenameInput = 'showColumnRenameInput',
+  hideColumnRenameInput = 'hideColumnRenameInput',
+  openCellContextMenu = 'openCellContextMenu',
+  openHeaderContextMenu = 'openHeaderContextMenu',
+  openHeaderDropdownMenu = 'openHeaderDropdownMenu',
+  closeContextMenu = 'closeContextMenu',
+  multipleCellsSelected = 'multipleCellsSelected',
+  gridSelectionCleared = 'gridSelectionCleared',
+  columnFreezeReset = 'columnFreezeReset',
+  columnFreezeChanged = 'columnFreezeChanged',
+  openSearch = 'openSearch',
+  closeSearch = 'closeSearch',
+}
+
+export const initialState: DatagridState = {
+  columns: [],
+  contextMenuData: DEFAULT_CONTEXT_MENU,
+  renameColumnInputData: DEFAULT_RENAME_INPUT_DATA,
+  gridSelection: EMPTY_GRID_SELECTION,
+  columnFreezeIndex: 0,
+  toggleSearch: false,
+  isResizeInProgress: false,
+};
+
+const typeToIconMap: Map<string, GridColumnIcon> = new Map([
+  [FieldType.number, GridColumnIcon.HeaderNumber],
+  [FieldType.string, GridColumnIcon.HeaderTextTemplate],
+  [FieldType.boolean, GridColumnIcon.HeaderBoolean],
+  [FieldType.time, GridColumnIcon.HeaderDate],
+  [FieldType.other, GridColumnIcon.HeaderReference],
+]);
+
+export const datagridReducer = (state: DatagridState, action: DatagridAction): DatagridState => {
```

Comment:
This code could be vastly reduced if you used redux toolkit where you can define the reducer and from the reducer actions and action types are automatically inferred .

https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/prometheus/querybuilder/components/PromQueryBuilderContainer.tsx#L66

Oh and you can use redux toolkit with useReducer (as shown in the link above)